azkaban-common/src/main/java/azkaban/executor/container/KubernetesContainerizedImpl.java

```diff
@@ -398,12 +402,78 @@ public void createContainer(final int executionId) throws ExecutorManagerExcepti
    */
   @Override
   public void deleteContainer(final int executionId) throws ExecutorManagerException {
-    deletePod(executionId);
+    try { // if pod deletion is not successful, the service deletion can still be handled
+      deletePod(executionId);
+    } catch (ExecutorManagerException e) {
+      logger.error("ExecId: {}, Unable to delete Pod in Kubernetes: {}", executionId,
+          e.getMessage());
+    }
     if (isServiceRequired()) {
       deleteService(executionId);
     }
   }
 
+  /**
+   * This method is used to delete invalid containers in batch
+   * @param containerValidity pod validity in milliseconds
+   * @throws ExecutorManagerException
+   */
+  @Override
+  public void deleteContainers(long containerValidity) throws ExecutorManagerException {
+    logger.info(String.format("Cleaning up containers older than %d days",
+        containerValidity / DAY_TO_MILLISECOND_CONVERSION_FACTOR));
+    final List<String> podsToDelete = getStalePods(containerValidity);
+    for (String podName : podsToDelete) {
+      logger.info(String.format("Deleting stale pod %s", podName));
+      final int execId = getExecutionId(podName);
+      if (execId > 0) {
+        deleteContainer(execId);
+      }
+    }
+  }
+
+  /**
+   * This method is used to fetch all stale pods in the current az cluster and namespace
+   * @param containerValidity pod validity in milliseconds
+   * @return List of stale pods
+   */
+  private List<String> getStalePods(final long containerValidity) {
+    List<String> stalePodList = new ArrayList<>();
+    try {
+      final String label = "cluster=" + this.clusterName;
+      V1PodList items= this.coreV1Api.listNamespacedPod(this.namespace, null,
+          null, null, null, label,
+          null, null, null, null);
+
+      // Get all pods whose age is older than Azkaban max flow running time (e.g. 10 days)
+      final Long validStartTimeStamp = System.currentTimeMillis() - containerValidity;
+      stalePodList =
+          items.getItems().stream()
+              .filter((pod) -> pod.getMetadata().getCreationTimestamp().getMillis() < validStartTimeStamp)
+              .map((pod) -> pod.getMetadata().getName()).collect(Collectors.toList());
+    } catch (ApiException e) {
+      logger.error(String.format("Unable to fetch stale pods in %s.", this.clusterName),
+          e.getResponseBody());
+    }
+    return stalePodList;
+  }
+
+  /**
+   * This method is used to extract execution id from pod mame
+   * @param podName
+   * @return execution id
+   */
+  private int getExecutionId(final String podName) {
+    int execId = -1;
+    try {
+      execId = Integer.valueOf(podName.replace(this.podPrefix + "-", "")
```

Comment:
Flow container pods are currently launched with the following label associated with them, e.g
`app=azkaban-exec-server,cluster=lasso,execution-id=execid-8606092,service=flow-lasso-8606092`
Checking these labels can serve as a good mechanism for:

1. Ensuring we are indeed deleting a flow-container pod and not a pod whose name accidentally happens to match the naming scheme `fc-xxx-yyy` 
2. Extracting the execution-id more directly as part of the label `execution-id`. 
3. Validating that the pod is for the current webserver with the label `cluster`. 

I would recommend matching based on the labels, let me know your thoughts on this.

Reply:
1  We are querying the pods under this web server and namespace, which should prevent irrelevant pods to be deleted
2. We can fetch execution-id string from label map, and extract the execid out of it. 
3. explained in 1